File: app/services/notification_service.py
"""Persistent, in-app notifications — no email/SMS/push infrastructure, no
scheduled job. Reminders are computed lazily, on each `GET /notifications`
call: the endpoint is polled every ~45s from the dashboard nav (see
static/admin/js/nav.js), so "due right now" is evaluated often enough to
feel live, and idempotent upserts mean polling never creates duplicates.
This keeps the whole feature dependency-free (no APScheduler/Celery —
CLAUDE.md §10: "every new dependency is a decision, not a default")."""

from datetime import datetime, timedelta, timezone

from app.core.database import get_global_settings, menu_log, notifications, purchase_schedule_items
from app.utils.meal_window import is_saturday, to_local


async def generate_due_notifications(now: datetime) -> None:
    global_settings = await get_global_settings()
    local_now = to_local(now, global_settings["local_timezone"])
    await _generate_prep_reminders(local_now, global_settings)
    await _generate_purchase_reminders(local_now, global_settings)


async def _generate_prep_reminders(local_now: datetime, global_settings: dict) -> None:
    today = local_now.date()
    meal_types = ["brunch"] if is_saturday(local_now) else ["breakfast", "lunch"]
    lead_minutes = global_settings["prep_lead_minutes"]

    for meal_type in meal_types:
        window = global_settings["meal_windows"].get(meal_type)
        if not window:
            continue

        has_menu_planned = await menu_log.count_documents(
            {
                "date": datetime.combine(today, datetime.min.time()),
                "meal_type": meal_type,
            },
            limit=1,
        )
        if not has_menu_planned:
            continue  # nothing planned for this meal — nothing to prep

        hour, minute = (int(part) for part in window["start"].split(":"))
        window_start = local_now.replace(hour=hour, minute=minute, second=0, microsecond=0)
        minutes_until_start = (window_start - local_now).total_seconds() / 60
        if not (0 <= minutes_until_start <= lead_minutes):
            continue

        await _upsert_notification(
            notif_type="prep_reminder",
            date_=today,
            meal_type=meal_type,
            title=f"Start prepping {meal_type}",
            message=f"{meal_type.capitalize()} service starts at {window['start']} — "
            f"about {int(minutes_until_start)} minute(s) from now.",
            visible_roles=["admin", "counter"],
        )


async def _generate_purchase_reminders(local_now: datetime, global_settings: dict) -> None:
    target_date = (local_now + timedelta(days=global_settings["purchase_lead_days"])).date()
    pending_count = await purchase_schedule_items.count_documents(
        {"date": datetime.combine(target_date, datetime.min.time()), "purchased": False}
    )
    if pending_count == 0:
        return

    await _upsert_notification(
        notif_type="purchase_due",
        date_=target_date,
        meal_type=None,
        title="Ingredient purchase due",
        message=f"{pending_count} item(s) still need buying for {target_date.isoformat()}.",
        visible_roles=["admin", "counter"],
    )


async def _upsert_notification(
    *, notif_type: str, date_, meal_type: str | None, title: str, message: str, visible_roles: list[str]
) -> None:
    now = datetime.now(timezone.utc)
    key = {"type": notif_type, "date": date_.isoformat(), "meal_type": meal_type}
    await notifications.update_one(
        key,
        {
            "$set": {"title": title, "message": message, "visible_roles": visible_roles, "updated_at": now},
            "$setOnInsert": {"created_at": now, "dismissed_by": []},
        },
        upsert=True,
    )


async def list_active_for_role(role: str, username: str) -> list[dict]:
    docs = (
        await notifications.find({"visible_roles": role, "dismissed_by": {"$ne": username}})
        .sort("created_at", -1)
        .to_list(length=100)
    )
    for doc in docs:
        doc["_id"] = str(doc["_id"])
    return docs


async def dismiss_notification(notification_id, username: str) -> bool:
    result = await notifications.update_one(
        {"_id": notification_id}, {"$addToSet": {"dismissed_by": username}}
    )
    return result.matched_count > 0

File: lib/services/notification_service.dart
import 'dart:convert';

import 'package:drift/drift.dart';

import '../core/app_mode.dart';
import '../core/errors.dart';
import '../core/role.dart';
import '../core/time/meal_window.dart';
import '../data/local/database.dart' hide Notification;
import '../data/local/mappers.dart';
import '../domain/ops.dart';
import '../domain/settings.dart';
import 'settings_service.dart';

/// Persistent in-app reminders — a port of v1
/// `app/services/notification_service.py` (PRD §6.5.2). No scheduled job:
/// [generateDue] is called on each `GET /notifications` poll, and the upserts
/// are idempotent so polling never duplicates.
class NotificationService {
  NotificationService(this._db, this._settings);

  final AppDatabase _db;
  final SettingsService _settings;

  Future<void> generateDue(DateTime nowUtc) async {
    final s = await _settings.read();
    final localNow = toLocal(nowUtc, s.localTimezone);
    await _generatePrepReminders(localNow, s);
    await _generatePurchaseReminders(localNow, s);
  }

  Future<void> _generatePrepReminders(
      DateTime localNow, SettingsSnapshot s) async {
    final today = DateTime.utc(localNow.year, localNow.month, localNow.day);
    final meals = isSaturday(localNow)
        ? const [MealType.brunch]
        : const [MealType.breakfast, MealType.lunch];

    for (final meal in meals) {
      final window = s.mealWindows[meal.wire];
      if (window == null) continue;

      final planned = await (_db.select(_db.menuEntries)
                ..where(
                    (e) => e.date.equals(today) & e.mealType.equals(meal.wire))
                ..limit(1))
              .getSingleOrNull() !=
          null;
      if (!planned) continue;

      final parts = window.start.split(':');
      final startLocal = DateTime(
        localNow.year,
        localNow.month,
        localNow.day,
        int.parse(parts[0]),
        int.parse(parts[1]),
      );
      final minutesUntil = startLocal.difference(localNow).inSeconds / 60.0;
      if (minutesUntil < 0 || minutesUntil > s.prepLeadMinutes) continue;

      await _upsert(
        type: 'prep_reminder',
        dateKey: _ymd(today),
        mealType: meal.wire,
        title: 'Start prepping ${meal.wire}',
        message:
            '${_capitalize(meal.wire)} service starts at ${window.start} — '
            'about ${minutesUntil.round()} minute(s) from now.',
        visibleRoles: const [Role.admin, Role.counter],
      );
    }
  }

  Future<void> _generatePurchaseReminders(
      DateTime localNow, SettingsSnapshot s) async {
    final target = DateTime.utc(localNow.year, localNow.month, localNow.day)
        .add(Duration(days: s.purchaseLeadDays));
    final pending = await (_db.select(_db.purchaseScheduleItems)
          ..where((p) => p.date.equals(target) & p.purchased.equals(false)))
        .get();
    if (pending.isEmpty) return;

    await _upsert(
      type: 'purchase_due',
      dateKey: _ymd(target),
      mealType: null,
      title: 'Ingredient purchase due',
      message:
          '${pending.length} item(s) still need buying for ${_ymd(target)}.',
      visibleRoles: const [Role.admin, Role.counter],
    );
  }

  /// Idempotent upsert keyed on (type, dateKey, mealType). Done by hand rather
  /// than a DB unique constraint because SQLite treats NULL mealType values as
  /// distinct, which would let `purchase_due` reminders pile up.
  Future<void> _upsert({
    required String type,
    required String dateKey,
    required String? mealType,
    required String title,
    required String message,
    required List<Role> visibleRoles,
  }) async {
    final now = DateTime.now().toUtc();
    final match = _db.select(_db.notifications)
      ..where((n) => n.type.equals(type) & n.dateKey.equals(dateKey));
    if (mealType == null) {
      match.where((n) => n.mealType.isNull());
    } else {
      match.where((n) => n.mealType.equals(mealType));
    }
    final existing = await match.getSingleOrNull();

    final rolesJson = jsonEncode(visibleRoles.map((r) => r.wire).toList());
    if (existing == null) {
      await _db.into(_db.notifications).insert(NotificationsCompanion.insert(
            type: type,
            dateKey: dateKey,
            mealType: Value(mealType),
            title: title,
            message: message,
            visibleRolesJson: rolesJson,
            createdAt: now,
            updatedAt: now,
          ));
    } else {
      await (_db.update(_db.notifications)
            ..where((n) => n.id.equals(existing.id)))
          .write(NotificationsCompanion(
        title: Value(title),
        message: Value(message),
        visibleRolesJson: Value(rolesJson),
        updatedAt: Value(now),
      ));
    }
  }

  /// Active + undismissed for this user, newest first.
  Future<List<AppNotification>> listFor(Role role, String username) async {
    final rows = await (_db.select(_db.notifications)
          ..orderBy([(n) => OrderingTerm.desc(n.createdAt)]))
        .get();
    return rows
        .where((n) {
          final roles =
              (jsonDecode(n.visibleRolesJson) as List<dynamic>).cast<String>();
          final dismissed =
              (jsonDecode(n.dismissedByJson) as List<dynamic>).cast<String>();
          return roles.contains(role.wire) && !dismissed.contains(username);
        })
        .map(notificationFromRow)
        .toList();
  }

  Future<void> dismiss(int id, String username) async {
    final row = await (_db.select(_db.notifications)
          ..where((n) => n.id.equals(id)))
        .getSingleOrNull();
    if (row == null) throw const NotFoundException('Notification not found.');
    final dismissed = (jsonDecode(row.dismissedByJson) as List<dynamic>)
        .cast<String>()
        .toSet()
      ..add(username);
    await (_db.update(_db.notifications)..where((n) => n.id.equals(id))).write(
      NotificationsCompanion(
          dismissedByJson: Value(jsonEncode(dismissed.toList()))),
    );
  }

  String _ymd(DateTime d) =>
      '${d.year.toString().padLeft(4, '0')}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';
  String _capitalize(String s) =>
      s.isEmpty ? s : s[0].toUpperCase() + s.substring(1);
}
